### ISSUE TITLE
XGBoost.py

### ISSUE THREAD

**Issue by GriMacez** [GriMacez/-#3]
import numpy as np
import pandas as pd
import xgboost as xgb
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error, mean_absolute_error
from scipy.fftpack import fft
from scipy.interpolate import make_interp_spline
import matplotlib.dates as mdates

# 固定随机种子，确保结果可复现
np.random.seed(42)

### 1. 数据加载（10分钟粒度，线性插值补全）
def load_weather_data(file_path):
    """加载10分钟粒度数据，处理日期并插值补全缺失值"""
    df = pd.read_csv(file_path)
    # 重命名列（适配常见气象数据集格式）
    df = df.rename(columns={
        'DateTime': 'date',
        'T (degC)': 'temperature',
        'rh (%)': 'humidity',
        'p (mbar)': 'pressure'
    })
    # 解析日期（兼容多种格式）
    try:
        df['date'] = pd.to_datetime(df['date'], format='%d.%m.%Y %H:%M:%S')
    except ValueError:
        df['date'] = pd.to_datetime(df['date'])  # 自动解析
    # 按时间排序（确保序列连续性）
    df = df.sort_values('date').reset_index(drop=True)
    # 线性插值补全（高频数据更适合插值）
    df[['temperature', 'humidity', 'pressure']] = df[['temperature', 'humidity', 'pressure']].interpolate(
        method='linear')
    return df

### 2. 特征工程（预测下一个10分钟，增强时间/波动特征）
def create_features(df, target_col='temperature', look_back=24, forecast_step=1):
    """
    look_back: 过去24个10分钟（4小时）的滞后特征
    forecast_step: 预测未来1个10分钟（统一任务）
    """
    df = df.copy()

    ### （1）细粒度时间特征 ###
    df['hour'] = df['date'].dt.hour  # 0-23小时
    df['minute'] = df['date'].dt.minute  # 0-59分钟
    df['day_of_week'] = df['date'].dt.dayofweek  # 周一0，周日6
    df['is_weekend'] = (df['day_of_week'] >= 5).astype(int)  # 周末标记
    # 时间段划分（凌晨0-6，上午6-12，下午12-18，晚上18-24）
    df['time_period'] = pd.cut(
        df['hour'],
        bins=[0, 6, 12, 18, 24],
        labels=[0, 1, 2, 3],
        include_lowest=True
    ).astype(int)

    ### （2）滞后特征（捕捉4小时内关联） ###
    for i in range(1, look_back + 1):
        for col in ['temperature', 'humidity', 'pressure']:
            df[f'{col}_lag_{i}'] = df[col].shift(i)  # 滞后i个10分钟

    ### （3）滚动统计（捕捉长周期趋势） ###
    windows = [look_back, look_back * 2, look_back * 3]  # 4h, 8h, 12h
    for window in windows:
        for col in ['temperature', 'humidity', 'pressure']:
            df[f'{col}_rolling_mean_{window}'] = df[col].rolling(window=window).mean()
            df[f'{col}_rolling_std_{window}'] = df[col].rolling(window=window).std()
            df[f'{col}_rolling_min_{window}'] = df[col].rolling(window=window).min()
            df[f'{col}_rolling_max_{window}'] = df[col].rolling(window=window).max()

    ### （4）傅里叶周期特征（捕捉日/周周期） ###
    def add_fft_feature(data, col, period, n_components=5):
        """提取傅里叶变换的前n_components个成分"""
        fft_vals = fft(data[col].values)
        fft_real = np.real(fft_vals[:n_components])
        fft_imag = np.imag(fft_vals[:n_components])
        for i in range(n_components):
            data[f'{col}_fft_real_{i + 1}'] = fft_real[i]
            data[f'{col}_fft_imag_{i + 1}'] = fft_imag[i]
        return data

    df = add_fft_feature(df, 'temperature', period=144, n_components=5)  # 1天周期
    df = add_fft_feature(df, 'temperature', period=1008, n_components=5)  # 1周周期

    ### （5）目标变量：下一个10分钟的温度 ###
    df[f'{target_col}_forecast'] = df[target_col].shift(-forecast_step)

    # 删除含NaN的行
    df = df.dropna().reset_index(drop=True)
    return df

### 3. 模型训练（XGBoost）
def train_xgb_model(X_train, y_train, X_val, y_val, feature_names):
    """训练XGBoost模型"""
    dtrain = xgb.DMatrix(X_train, label=y_train, feature_names=feature_names)
    dval = xgb.DMatrix(X_val, label=y_val, feature_names=feature_names)

    # 参数调整
    params = {
        'objective': 'reg:squarederror',
        'eval_metric': 'rmse',
        'max_depth': 10,
        'learning_rate': 0.01,
        'subsample': 0.7,
        'colsample_bytree': 0.7,
        'n_estimators': 1000,
        'gamma': 0.1,
        'reg_alpha': 0.1,
        'reg_lambda': 0.1,
        'min_child_weight': 5,
        'verbosity': 1,
        'nthread': -1
    }

    watchlist = [(dtrain, 'train'), (dval, 'val')]
    model = xgb.train(
        params,
        dtrain,
        num_boost_round=3000,
        evals=watchlist,
        early_stopping_rounds=50,
        verbose_eval=50
    )
    return model

def recursive_predict(model, df_features, feature_names, look_back=24, forecast_steps=72, noise_factor=1.0):
    """递归预测未来温度，确保特征动态更新"""
    print("预测时特征列数：", len(feature_names))
    cols = ['temperature', 'humidity', 'pressure']
    historical = {col: df_features[col].iloc[-look_back:].tolist() for col in cols}

    # 计算历史标准差
    hum_std = df_features['humidity'].iloc[-look_back:].rolling(look_back).std().iloc[-1]
    pres_std = df_features['pressure'].iloc[-look_back:].rolling(look_back).std().iloc[-1]

    last_features = df_features[feature_names].iloc[-1:].copy()
    last_date = df_features['date'].iloc[-1]
    future_dates = pd.date_range(
        start=last_date + pd.Timedelta(minutes=10),
        periods=forecast_steps,
        freq='10T'
    )
    future_preds = []

    windows = [look_back, look_back * 2, look_back * 3]  # 4h, 8h, 12h

    for i in range(forecast_steps):
        # 预测并保存结果
        dtest = xgb.DMatrix(last_features, feature_names=feature_names)
        try:
            pred_temp = model.predict(dtest)[0]
        except Exception as e:
            print(f"预测第 {i + 1}/{forecast_steps} 步时出错: {e}")
            print("当前特征值:", last_features[['temperature_lag_1', 'temperature_rolling_mean_24']])
            break

        future_preds.append(pred_temp)

        # 更新湿度和气压（添加适当的随机波动）
        new_hum = historical['humidity'][-1] + np.random.normal(0, hum_std * 0.2)
        new_pres = historical['pressure'][-1] + np.random.normal(0, pres_std * 0.2)

        # 更新历史数据
        for col, new_val in zip(cols, [pred_temp, new_hum, new_pres]):
            historical[col].append(new_val)
            historical[col] = historical[col][1:]  # 保持固定长度

        # 强制更新所有滞后特征，加入适当扰动
        for col in cols:
            for i in range(1, look_back + 1):
                col_name = f'{col}_lag_{i}'
                if col_name in feature_names:
                    last_features[col_name] = historical[col][-i] + np.random.normal(0, noise_factor)

        # 强制更新所有滚动特征
        for window in windows:
            for col in cols:
                for stat in ['mean', 'std', 'min', 'max']:
                    col_name = f'{col}_rolling_{stat}_{window}'
                    if col_name in feature_names:
                        window_vals = historical[col][-window:]
                        if stat == 'mean':
                            last_features[col_name] = np.mean(window_vals)
                        elif stat == 'std':
                            last_features[col_name] = np.std(window_vals)
                        elif stat == 'min':
                            last_features[col_name] = np.min(window_vals)
                        elif stat == 'max':
                            last_features[col_name] = np.max(window_vals)

    # 如果预测提前中断，截取相应长度的 future_dates
    future_dates = future_dates[:len(future_preds)]

    return future_dates, future_preds

### 特征重要性可视化函数
def plot_feature_importance(model, feature_names, top_n=20, save_path='feature_importance.png'):
    importances = model.get_score(importance_type='gain')
    df_imp = pd.DataFrame({
        'feature': list(importances.keys()),
        'importance': list(importances.values())
    })
    df_imp = df_imp.sort_values('importance', ascending=False).head(top_n)

    plt.figure(figsize=(10, 8))
    bars = plt.barh(df_imp['feature'], df_imp['importance'], color='#1f77b4')
    plt.xlabel('Feature Importance (Gain)', fontsize=12)
    plt.ylabel('Feature', fontsize=12)
    plt.gca().invert_yaxis()

    for bar, val in zip(bars, df_imp['importance']):
        width = bar.get_width()
        plt.text(width + 5000,
                 bar.get_y() + bar.get_height() / 2,
                 f'{int(val):,}',
                 va='center', fontsize=10, color='darkred')

    plt.yticks(fontsize=9, rotation=30)
    plt.title(f'XGBoost 特征重要性（增益，前{top_n}）', fontsize=14)
    plt.tight_layout()
    plt.savefig(save_path)
    plt.close()

def replace_invalid_values(series):
    """替换Series中的NaN和inf值"""
    series = series.copy()
    series = series.replace([np.inf, -np.inf], np.nan)
    series = series.fillna(method='ffill').fillna(method='bfill')  # 前后向填充
    return series

def main():
    # ------------------- 数据加载 -------------------
    # 替换为实际数据文件路径，需确保文件有 'Date Time', 'T (degC)', 'rh (%)', 'p (mbar)' 列
    df = load_weather_data('../data.csv')
    print(f"✅ 数据加载完成 | 共 {len(df)} 条 | 时间范围: {df['date'].min()} ~ {df['date'].max()}")

    # ------------------- 特征工程 -------------------
    target_col = 'temperature'
    look_back = 24  # 过去4小时（24个10分钟）
    forecast_step = 1  # 预测下一个10分钟（统一任务）
    df_features = create_features(df, target_col, look_back, forecast_step)
    print(f"✅ 特征工程完成 | 特征数: {len(df_features.columns)} | 有效样本: {len(df_features)}")

    # ------------------- 数据集划分 -------------------
    X = df_features.drop(['date', f'{target_col}_forecast'], axis=1)
    y = df_features[f'{target_col}_forecast']

    train_size = int(0.8 * len(X))
    val_size = int(0.1 * len(X))
    X_train, X_val, X_test = X.iloc[:train_size], X.iloc[train_size:train_size + val_size], X.iloc[
                                                                                            train_size + val_size:]
    y_train, y_val, y_test = y.iloc[:train_size], y.iloc[train_size:train_size + val_size], y.iloc[
                                                                                            train_size + val_size:]

    print(f"✅ 数据集划分 | 训练集: {len(X_train)} | 验证集: {len(X_val)} | 测试集: {len(X_test)}")

    # ------------------- 模型训练 -------------------
    feature_names = X.columns.tolist()
    model = train_xgb_model(X_train, y_train, X_val, y_val, feature_names)  # 训练模型

    # ------------------- 测试集评估 -------------------
    dtest = xgb.DMatrix(X_test, feature_names=feature_names)
    y_pred_test = model.predict(dtest)
    rmse = np.sqrt(mean_squared_error(y_test, y_pred_test))
    mae = mean_absolute_error(y_test, y_pred_test)
    print(f"📊 测试集性能 | RMSE: {rmse:.2f} | MAE: {mae:.2f}")

    # ------------------- 测试集可视化 -------------------
    try:
        plt.rcParams['font.sans-serif'] = ['SimHei', 'WenQuanYi Micro Hei', 'Heiti TC']  # 增加更多字体选项
        plt.rcParams['axes.unicode_minus'] = False
    except Exception as e:
        print(f"字体设置失败: {e}，使用默认字体")

    plot_start = max(0, len(df_features) - 200)
    test_dates = df_features['date'].iloc[X_test.index]

    plt.figure(figsize=(16, 8))
    plt.plot(test_dates, y_test, label='实际温度', color='blue', alpha=0.8)
    plt.plot(test_dates, y_pred_test, label='预测温度', color='red', linestyle='--', alpha=0.8)
    plt.title('10分钟粒度温度预测对比（测试集）')
    plt.xlabel('日期')
    plt.ylabel('温度 (°C)')
    plt.legend()
    plt.grid(True)
    plt.tight_layout()
    plt.savefig('test_prediction.png')
    plt.close()
    print("✅ 测试集预测图保存: test_prediction.png")

    # ------------------- 特征重要性可视化 -------------------
    plot_feature_importance(model, feature_names, top_n=20, save_path='feature_importance.png')
    print("✅ 特征重要性图保存: feature_importance.png")

    # ------------------- 未来预测（720分钟=12小时） -------------------
    forecast_steps = 72  # 72个10分钟=720分钟=12小时
    future_dates, future_preds = recursive_predict(model, df_features, feature_names, look_back, forecast_steps)

    # 计算预测温度的波动范围
    pred_series = pd.Series(future_preds)
    rolling_std = pred_series.rolling(window=5, min_periods=1).std()
    upper_bound = pred_series + rolling_std
    lower_bound = pred_series - rolling_std

    # 检查并处理可能的NaN或inf值
    upper_bound = replace_invalid_values(upper_bound)
    lower_bound = replace_invalid_values(lower_bound)

    # 可视化未来预测 + 历史200个10分钟数据 + 温度范围
    plt.figure(figsize=(16, 8))
    hist_start = max(0, len(df) - 200)
    plt.plot(df['date'].iloc[hist_start:], df['temperature'].iloc[hist_start:],
             label='历史温度（最后200个10分钟）', color='blue', alpha=0.75, linewidth=2)

    # 平滑绘制未来预测曲线
    if len(future_preds) > 10:
        future_x = mdates.date2num(future_dates)

        # 确保预测值也没有无效值
        pred_series_clean = replace_invalid_values(pred_series)

        spline = make_interp_spline(future_x, pred_series_clean, k=3)
        x_smooth = np.linspace(future_x.min(), future_x.max(), 300)
        y_smooth = spline(x_smooth)

        # 平滑上下界曲线
        upper_spline = make_interp_spline(future_x, upper_bound, k=3)
        lower_spline = make_interp_spline(future_x, lower_bound, k=3)
        upper_smooth = upper_spline(x_smooth)
        lower_smooth = lower_spline(x_smooth)

        # 绘制平滑曲线
        plt.plot(
            mdates.num2date(x_smooth),
            y_smooth,
            label='未来温度预测（平滑）',
            color='green',
            linewidth=2.5,
            alpha=0.85
        )
    else:
        # 数据点不足时不进行平滑处理
        plt.plot(
            future_dates,
            future_preds,
            label='未来温度预测',
            color='green',
            linewidth=2.5,
            alpha=0.85
        )
        upper_smooth = upper_bound
        lower_smooth = lower_bound
        x_smooth = future_x = mdates.date2num(future_dates)

    # 绘制温度波动范围
    plt.fill_between(
        mdates.num2date(x_smooth),
        lower_smooth,
        upper_smooth,
        color='green',
        alpha=0.2,
        label='温度波动范围'
    )

    # 设置x轴格式为“月-日 时:分”
    plt.gca().xaxis.set_major_formatter(mdates.DateFormatter('%m-%d %H:%M'))
    plt.gca().xaxis.set_major_locator(mdates.MinuteLocator(interval=60))  # 每小时一个刻度
    plt.gcf().autofmt_xdate()  # 自动旋转日期标签，避免重叠

    plt.title('未来12小时温度预测（10分钟粒度）')
    plt.xlabel('时间')
    plt.ylabel('温度 (°C)')
    plt.legend()
    plt.grid(True)
    plt.tight_layout()

    # 保存图像
    plt.savefig('future_prediction.png', dpi=300)
    print("✅ 未来预测图保存为 future_prediction.png")
    plt.show()
    plt.close()

if __name__ == "__main__":
    main()
